src/mpid/ch4/netmod/include/netmod.h

```diff
@@ -51,6 +51,9 @@ typedef int (*MPIDI_NM_am_isend_reply_t) (MPIR_Context_id_t context_id, int src_
 typedef size_t(*MPIDI_NM_am_hdr_max_sz_t) (void);
 typedef size_t(*MPIDI_NM_am_eager_limit_t) (void);
 typedef size_t(*MPIDI_NM_am_eager_buf_limit_t) (void);
+typedef bool(*MPIDI_NM_am_check_eager_t) (MPI_Aint am_hdr_sz, MPI_Aint data_sz, const void *data,
+                                          MPI_Count count, MPI_Datatype datatype,
+                                          MPIR_Request * sreq);
```

Comment:
I think at ch4-level we have 3 am mode now: `eager`, `rendezvous`, and `zcopy`. The `zcopy` is not the same as `eager` because, the ch4-layer send side need set the flag -- similar to the flag we set for `ssend` so the ch4-layer target side will know not to allocate space for unexpected message and call netmod/shmmod zcopy when recv is posted. It is distinctly different from eager just from ch4-layer perspective. This is different from pipeline where the handling is completely hidden from ch4.

So this function really should return a 3-way choice. If you just return eager, then you have to use some other mechanism later to distinguish the handling between zcopy and normal eager, which brings more complexity, right?

If we return a enum choice, then I think this method really should be named something like `MPIDI_NM_am_check_method`.

Reply:
I am also concerned that we are passing in `data_sz`, which is redundant from `buf, count, datatype`. I understand your reason to avoid extra macro call, but from API point of view, the obvious redundancy is un-intuitive. I don't have any better solution now, so I am not suggesting your change it, just something to think about.

Reply:
@yfguo Would you like to reconsider the idea of having the query function return an opaque extended header to ch4? By doing that, we can remove all the logic in `MPIDI_NM_am_isend` that deals with `zcopy` vs `eager`, both on the send side and target side. All the zcopy decision and preparations are made in the query function and opaquely passed by ch4. When ch4 being called in callback, if it is a `zcopy`, it can simply store the opaque info with the request. This opaque blob will be passed back to netmod with `MPIDI_NM_am_zcopy_start`.

Essentially we are making `zcopy` entirely a ch4 function, removing all the logics from the netmod's am path. Netmod just need supply the query function and zcopy_start to implement a zcopy.